src/lib/snark_work_lib/partitioned.ml

```diff
@@ -0,0 +1,599 @@
+                match stmt with
+                | Ok stmt ->
+                    stmt
+                | Error e ->
+                    failwithf
+                      "Failed to construct a statement from  zkapp merge \
+                       command %s"
+                      (Error.to_string_hum e) () )
+        end
+      end]
+
+      [%%define_locally
+      Stable.Latest.(t_of_sexp, sexp_of_t, to_yojson, of_yojson, map, statement)]
+    end
+
+    [%%versioned
+    module Stable = struct
+      [@@@no_toplevel_latest_type]
+
+      module V1 = struct
+        type t =
+          ( Transaction_snark.Zkapp_command_segment.Witness.Stable.V1.t
+          , Ledger_proof.Stable.V2.t )
+          Poly.Stable.V1.t
+        [@@deriving sexp, yojson]
+
+        let statement : t -> Transaction_snark.Statement.t = Poly.statement
+
+        let to_latest = Fn.id
+      end
+    end]
+
+    type t =
+      ( Transaction_snark.Zkapp_command_segment.Witness.t
+      , Ledger_proof.Cached.t )
+      Poly.t
+
+    let read_all_proofs_from_disk : t -> Stable.Latest.t =
+      Poly.map
+        ~f_witness:
+          Transaction_snark.Zkapp_command_segment.Witness
+          .read_all_proofs_from_disk
+        ~f_proof:Ledger_proof.Cached.read_proof_from_disk
+
+    let write_all_proofs_to_disk ~(proof_cache_db : Proof_cache_tag.cache_db) :
+        Stable.Latest.t -> t =
+      Poly.map
+        ~f_witness:
+          (Transaction_snark.Zkapp_command_segment.Witness
+           .write_all_proofs_to_disk ~proof_cache_db )
+        ~f_proof:(Ledger_proof.Cached.write_proof_to_disk ~proof_cache_db)
+  end
+
+  module Poly = struct
+    [%%versioned
+    module Stable = struct
+      module V1 = struct
+        type 'spec t =
+          { spec : 'spec
+          ; pairing : Pairing.Sub_zkapp.Stable.V1.t
+          ; job_id : ID.Stable.V1.t
+          ; common : Spec_common.Stable.V1.t
+          }
+        [@@deriving sexp, yojson]
+
+        let map ~(f_spec : 'a -> 'b) (t : 'a t) : 'b t =
+          { t with spec = f_spec t.spec }
+      end
+    end]
+
+    [%%define_locally
+    Stable.Latest.(t_of_sexp, sexp_of_t, to_yojson, of_yojson, map)]
+  end
+
+  [%%versioned
+  module Stable = struct
+    [@@@no_toplevel_latest_type]
+
+    module V1 = struct
+      type t = Spec.Stable.V1.t Poly.Stable.V1.t [@@deriving sexp, yojson]
+
+      let to_latest = Fn.id
+    end
+  end]
+
+  type t = Spec.t Poly.t
+
+  let read_all_proofs_from_disk : t -> Stable.Latest.t =
+    Poly.map ~f_spec:Spec.read_all_proofs_from_disk
+
+  let write_all_proofs_to_disk ~(proof_cache_db : Proof_cache_tag.cache_db) :
+      Stable.Latest.t -> t =
+    Poly.map ~f_spec:(Spec.write_all_proofs_to_disk ~proof_cache_db)
+end
+
+module Spec = struct
+  module Poly = struct
+    [%%versioned
+    module Stable = struct
+      module V1 = struct
+        type ( 'witness
+             , 'zkapp_command_segment_witness
+             , 'ledger_proof
+             , 'metric )
```

Comment:
It seems wrong to call this parameter `metric`, because it's later substituted with `Proof_with_metric` which is like not entirely a metric, but rather a result of execution. I'd suggest a generic `data` to be used here, if there is no better idea.

Actually, after thinking about it for another minute, it seems like we could drop `'metric` from the spec entirely once we move forward with the decision to not have any backwards-compatibility in RPC. Then `Old` constructor will disappear. And spec is a spec, and result is a pair of spec and proof-with-metric.